Feature: Add support for generating Windows Store listing data.

orudge opened the pull request:
Microsoft flagged up our Windows Store listing as lacking translations for the many other languages that we say we support. (Historically I just copied the English data to these other stores). To ensure we don't get kicked out of the Store, I've submitted machine-translated versions for now for all the supported languages, but it would be preferable if we can have proper translations. Luckily, this problem has already been solved for Steam, so this is my attempt to do so for the Windows Store too.

The Windows Store uses a .csv file to allow import and export of listing data, which is a bit of a pain. To make my life easier, I've written a small C# program (tested in .NET 8) to take the CSV template and generate the multiple language listings. It also enumerates the screenshots, so we should have a better selection on the Store than the single ancient screenshot I uploaded years ago. :)

In the store listing, I've basically just reused the existing Steam fields for the Windows Store. I still need to test that the generated .csv and screenshots import correctly into the Store (but can't do this until MS has approved the submission I made earlier today) - there are some length limitations on some fields, so I may need to make some tweaks yet.

There are 22 languages supported by the Store that are not currently in the repository; I added stubs for them (based on 8686de0e7235), but if I need to do something else to get those languages to show up in the translator, let me know.

orudge committed:
Feature: Add support for generating Windows Store listing data.

orudge committed:
Add: Stubs for languages supported by the Windows Store.

TrueBrain commented:
Mostly curious, any specific reason this being a C# application? From what I can tell, this can be done with a lot less lines in a simple Python script, but I might be missing some detail that got lost in the C# boilerplate :D

Not that I mind the C#, but we already have some Python in this repo; keeping things more the same makes things easier. But again, I just might be missing something in the C# code :)

orudge commented:
Mainly because I know C# and I don't really know Python, so it would have taken me far longer to achieve the same in Python. :D The code should run on any platform .NET runs on (it's not Windows-specific). At present the code will just be run manually by me (automating Windows Store uploads may be possible, but is non-trivial).

TrueBrain commented:
All stores currently are updated manually, so that is perfectly fine. Automating that is tricky, so I didn't bother with the other stores. Translations also rarely change.

orudge commented:
So despite there being 30 `DesktopScreenshot` lines in the CSV, the web site will only accept 10. But there's not much point in uploading them every time we do an app update, so I've uploaded them manually for now and added the references to the template. Also added in the trailer and a couple of other bits of artwork.

This should be ready to go, then hopefully we can get translations in for the other languages and actually use this.

TrueBrain commented on winstore/Program.cs:
```diff
@@ -0,0 +1,271 @@
+
+                if (!File.Exists(path))
+                {
+                    // TODO: For now, substitute the English version.
```
The downside of `TODO`s: who is going to pick it up when?

So maybe the question .. is it actually a TODO?

orudge replied:
Well, the hope is that once the translations are in EINTS, people translate them, and we no longer need this. :) So I suppose it depends on how long the translations take.

TrueBrain replied:
~~Wouldn't it always be better to use the english sentence for those translations that do not have a translated string? That is more why I meant if this is actually a TODO. As I don't think you want to do it "for now", but in all cases where the translation is without an actual translation :)~~

TrueBrain replied:
Owh, I was misreading the if-statement .. now you created all the language files, is this code that is actually executed? I think it would be better if the tool crashes (or exits with a warning) at this point, not? As that means you didn't create a placeholder file where one is expected?

In other words, eints is not going to create any new files when things are translated; it will only update the ones that currently exist.

orudge replied:
Sorry, forgot about this over Christmas! In theory, yes, you're correct - we shouldn't have any more files appearing, unless we add more languages in the Windows Store. So that could well be removed, though maybe having it display a warning would be safest.

orudge replied:
An exception will now occur if the language file is missing.

TrueBrain commented on winstore/Program.cs:
```diff
@@ -0,0 +1,271 @@
+            new Language("ukrainian", "uk-ua"),
+            new Language("vietnamese", "vi-vn"),
+
+            // The following languages are supported by the Store, but not currently translated
```
With this PR, that is no longer true, is it? So maybe already remove the comment, I would think.

orudge replied:
I'll remove the comment; they were the languages that didn't have .txt files at the time I wrote it.

TrueBrain commented on winstore/Program.cs:
```diff
@@ -0,0 +1,271 @@
+            new Language("czech", "cs-cz"),
+            new Language("danish", "da-dk"),
+            new Language("dutch", "nl-nl"),
+            new Language("finnish", "fi-fi"), // Not currently selected for OpenTTD on the Windows Store
```
If that comment is true, why is it in here? Or might the comment become untrue at some point? Or should we disable the line in total? :)

orudge replied:
It's more a "todo" comment for me; once we have translations running through this, I'll add it to the WinStore listing. I'll remove the comment now.

TrueBrain reviewed (commented):
Small nitpicks, mostly to make sure we are not going to say: "huh?!" in a year or so :)

I tried to remember whether we could just merge this or not, but I really can't remember. So when you looked at my comments, we are just going to YOLO this. But it might be I need to run some magic command to get the languages visible in eints ... one problem at the time :D

TrueBrain commented on lang/welsh.txt:
```diff
@@ -0,0 +1 @@
+##plural 0
```
Are all these languages actually `plural 0`? Or is it just too much copy/paste?

orudge replied:
I have no idea, I didn't realise the values were meaningful in the original stubs (I assumed they'd be overwritten by EINTS).

I'll check them against the OpenTTD versions. :)

TrueBrain replied:
Eints can do a lot of things, but I don't think this is one of them :D But honestly, I don't know! All I saw was that the original commit already had them correct :P Better safe than sorry! Tnx for fixing them :)

TrueBrain commented on winstore-template.csv:
```diff
@@ -0,0 +1,430 @@
+- GUI improvements, scalable font
+- New Help and Manuals option, improved default settings
+
+Check out our blogs on openttd.org for full details."
```
I am a bit slow, but I missed this entry in my first review.

I would very much prefer if we don't have changelogs hardcoded in this repository. Do we actually need to have it like this? As this feels like asking for trouble :D (and another step in the release pipeline). Is there any way around this, or a way to avoid it?

(to be clear, I am not going to keep back this PR because of this; just want to challenge you if we actually want/need this)

Edit: it is also not translated; so maybe just put an URL in there to the changelog, or website? Not sure .. just thinking out loud, mostly.

orudge replied:
I suppose I can just put it in a .txt file that's not committed to the repository and read it in from there, perhaps. I need to manually generate these release notes for the Store, but there's a character limitation on this field so I can't just use the changelog from OpenTTD as-is.

I could also put in a "Visit https://url/ for the latest changes" or even the awful "Bug fixes and enhancements" that every mobile app now has, but I prefer to actually have some details for users to see. :)

And yes, it's not translated, but neither is changelog.txt in OpenTTD of course.

orudge replied:
I've removed the release notes from the repository and have added a placeholder text file, which I'll update manually when doing a release. If we ever get to the point of automating Windows Store releases, then it may well be replaced with a URL, but while I'm still doing it manually, I'm happy to maintain the release notes myself.

TrueBrain commented on winstore-template.csv:
```diff
@@ -0,0 +1,430 @@
+VoiceTitle,7,Text,,
+ShortDescription,8,Text,,$STR_ABSTRACT$
+DevStudio,9,Text,,OpenTTD Developers
+CopyrightTrademarkInformation,12,Text,,Copyright © OpenTTD Developers 2002-2024
```
Maybe automate the `2024` while doing all this scripting. You will forget to update this every single year, I can promise you :P

orudge replied:
Yes, if I'm taking the changelog out of the .csv template then I should probably do that too!

orudge replied:
This will now default to the current year as suggested.
